fassst.py

```diff
@@ -60,18 +80,35 @@ def is_pure(node):
         if not is_pure(node.iter):
             return node
 
+        loop_id = self.loop_id
+        self.loop_id_stack.append(loop_id)
+        self.loop_id += 1
+
         new_body = []
-        for x in iterator_elements(node.iter):
-            # FIXME: this should probably use the location the expression
-            # actually appeared in (the range(n) call or the element in a
-            # literal list/tuple).
-            ast.fix_missing_locations(x)
+        for i, x in enumerate(iterator_elements(node.iter)):
             new_body.append(
                 ast.copy_location(ast.Assign(targets=[node.target], value=x), node)
             )
-            new_body.extend(node.body)
+            for n in node.body:
+                new_n = self.visit(n)
+                if isinstance(new_n, list):
+                    new_body.extend(new_n)
+                else:
+                    new_body.append(new_n)
+            new_body.append(make_placeholder("iteration_end", loop_id, i))
+        new_body.append(make_placeholder("loop_end", loop_id))
+
+        self.loop_id_stack.pop()
         return new_body
 
+    def visit_Break(self, node):
```

Comment:
Hmm I think visit_Break and visit_Continue break with while loops, both top-level and nested for-loops. We could solve this kinda nicely by creating another `ast.NodeTransformer` that we use to `visit` above.

Reply:
Yeah good catch 👍